netaddr_test.go

```diff
@@ -272,6 +273,135 @@ func TestIs4In6(t *testing.T) {
 	}
 }
 
+func TestIPPrefix(t *testing.T) {
+	tests := []struct {
+		prefix      string
+		ip          IP
+		bits        uint8
+		contains    []IP
+		notContains []IP
+	}{
+		{
+			prefix:      "192.168.0.0/24",
+			ip:          mustIP("192.168.0.0"),
+			bits:        24,
+			contains:    mustIPs("192.168.0.1", "192.168.0.55"),
+			notContains: mustIPs("192.168.1.1", "1.1.1.1"),
+		},
+		{
+			prefix:      "192.168.1.1/32",
+			ip:          mustIP("192.168.1.1"),
+			bits:        32,
+			contains:    mustIPs("192.168.1.1"),
+			notContains: mustIPs("192.168.1.2"),
+		},
+		{
+			prefix:      "2001:db8::/96",
+			ip:          mustIP("2001:db8::"),
+			bits:        96,
+			contains:    mustIPs("2001:db8::aaaa:bbbb", "2001:db8::1"),
+			notContains: mustIPs("2001:db8::1:aaaa:bbbb", "2001:db9::"),
+		},
+		{
+			prefix:   "0.0.0.0/0",
+			ip:       mustIP("0.0.0.0"),
+			bits:     0,
+			contains: mustIPs("192.168.0.1", "1.1.1.1"),
+		},
+		{
+			prefix:   "::/0",
+			ip:       mustIP("::"),
+			bits:     0,
+			contains: mustIPs("::1", "2001:db8::1"),
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.prefix, func(t *testing.T) {
+			prefix, err := ParseIPPrefix(test.prefix)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if prefix.IP != test.ip {
+				t.Errorf("IP=%s, want %s", prefix.IP, test.ip)
+			}
+			if prefix.Bits != test.bits {
+				t.Errorf("bits=%d, want %d", prefix.Bits, test.bits)
+			}
+			for _, ip := range test.contains {
+				if !prefix.Contains(ip) {
+					t.Errorf("does not contain %s", ip)
+				}
+			}
+			for _, ip := range test.notContains {
+				if prefix.Contains(ip) {
+					t.Errorf("contains %s", ip)
+				}
+			}
+			if got := prefix.String(); got != test.prefix {
+				t.Errorf("prefix.String()=%q, want %q", got, test.prefix)
+			}
+		})
+	}
+}
+
+func TestParseIPPrefixAllocs(t *testing.T) {
+	ip := "192.168.1.0"
```

Comment:
Nit: test IPv6 allocations as well for completeness?

Reply:
seems entirely unnecessary given the design of the test, but sure. done.

Reply:
Thanks! For my own curiosity: what makes it unnecessary?

I see that ParseIPPrefix uses a single code path for v4/v6, but I do see a slight divergence for ParseIP due to the different underlying types. I figured it would make sense to verify both are zero allocation, but perhaps I misunderstand?

Reply:
The test as written runs both ParseIP and ParseIPPrefix. It subtracts the allocs of the former from the latter. So if ParseIP goes and generates N more allocs for IPv6, it makes no difference to the accounting here.

Besides the ParseIP which we are explicitly subtracting, there are no divergent codepaths for v4 c.f. v6 so adding a v6 address does not increase effective coverage.

This is an example of where a code coverage tool would generate an incorrect report, because it can't see the test is annulling a lot of code paths by doing `allocs(ParseIPPrefix)-allocs(ParseIP)`.

Reply:
Understood, thanks for the follow up!